src/modules/commander/calibration_routines.cpp

```diff
@@ -233,7 +234,299 @@ int sphere_fit_least_squares(const float x[], const float y[], const float z[],
+
+		for (uint8_t i = 0; i < 9; i++) {
+			// compute JTJ
+			for (uint8_t j = 0; j < 9; j++) {
+				JTJ[i * 9 + j] += ellipsoid_jacob[i] * ellipsoid_jacob[j];
+				JTJ2[i * 9 + j] += ellipsoid_jacob[i] * ellipsoid_jacob[j]; //a backup JTJ for LM
+			}
+
+			JTFI[i] += ellipsoid_jacob[i] * residual;
+		}
+	}
+
+
+	//------------------------Levenberg-Marquardt-part-starts-here---------------------------------//
+	//refer: http://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm#Choice_of_damping_parameter
+	float fit1_params[9] = {*offset_x, *offset_y, *offset_z, *diag_x, *diag_y, *diag_z, *offdiag_x, *offdiag_y, *offdiag_z};
+	float fit2_params[9];
+	memcpy(fit2_params, fit1_params, sizeof(fit1_params));
+
+	for (uint8_t i = 0; i < 9; i++) {
+		JTJ[i * 9 + i] += _sphere_lambda;
+		JTJ2[i * 9 + i] += _sphere_lambda / lma_damping;
+	}
+
+
+	if (!mat_inverse(JTJ, JTJ, 9)) {
+		return -1;
+	}
+
+	if (!mat_inverse(JTJ2, JTJ2, 9)) {
+		return -1;
+	}
+
+
+
+	for (uint8_t row = 0; row < 9; row++) {
+		for (uint8_t col = 0; col < 9; col++) {
+			fit1_params[row] -= JTFI[col] * JTJ[row * 9 + col];
+			fit2_params[row] -= JTFI[col] * JTJ2[row * 9 + col];
+		}
+	}
+
+	//Calculate mean squared residuals
+	for (uint16_t k = 0; k < _samples_collected; k++) {
+		float A = (fit1_params[3]    * (x[k] - fit1_params[0])) + (fit1_params[6] * (y[k] - fit1_params[1])) + (fit1_params[7] *
+				(z[k] - fit1_params[2]));
+		float B = (fit1_params[6] * (x[k] - fit1_params[0])) + (fit1_params[4]   * (y[k] - fit1_params[1])) + (fit1_params[8] *
+				(z[k] - fit1_params[2]));
+		float C = (fit1_params[7] * (x[k] - fit1_params[0])) + (fit1_params[8] * (y[k] - fit1_params[1])) + (fit1_params[5]    *
+				(z[k] - fit1_params[2]));
+		float length = sqrtf(A * A + B * B + C * C);
+		residual = *sphere_radius - length;
+		fit1 += residual * residual;
+
+		A = (fit2_params[3]    * (x[k] - fit2_params[0])) + (fit2_params[6] * (y[k] - fit2_params[1])) + (fit2_params[7] *
+				(z[k] - fit2_params[2]));
+		B = (fit2_params[6] * (x[k] - fit2_params[0])) + (fit2_params[4]   * (y[k] - fit2_params[1])) + (fit2_params[8] *
+				(z[k] - fit2_params[2]));
+		C = (fit2_params[7] * (x[k] - fit2_params[0])) + (fit2_params[8] * (y[k] - fit2_params[1])) + (fit2_params[5]    *
+				(z[k] - fit2_params[2]));
+		length = sqrtf(A * A + B * B + C * C);
+		residual = *sphere_radius - length;
+		fit2 += residual * residual;
+	}
+
+	fit1 = sqrtf(fit1) / _samples_collected;
+	fit2 = sqrtf(fit2) / _samples_collected;
+
+	if (fit1 > _fitness && fit2 > _fitness) {
+		_sphere_lambda *= lma_damping;
+
+	} else if (fit2 < _fitness && fit2 < fit1) {
+		_sphere_lambda /= lma_damping;
+		memcpy(fit1_params, fit2_params, sizeof(fit1_params));
+		fitness = fit2;
+
+	} else if (fit1 < _fitness) {
+		fitness = fit1;
+	}
+
+	//--------------------Levenberg-Marquardt-part-ends-here--------------------------------//
+	if (!isnan(fitness) && fitness < _fitness) {
```

Comment:
should be PX4_ISNAN() because `isnan` is not well-defined in a portable code sense.